src/puppetlabs/puppetdb/pql/transform.clj

```diff
@@ -1,19 +1,44 @@
 (ns puppetlabs.puppetdb.pql.transform
   (:require [clojure.string :as str]
-            [puppetlabs.puppetdb.cheshire :as json]))
+            [puppetlabs.puppetdb.cheshire :as json]
+            [puppetlabs.puppetdb.zip :as zip]
+            [clojure.core.match :as cm]))
 
 (defn paging-clause?
   [v]
   (contains? #{"limit" "offset" "order_by"} (first v)))
 
+(defn strip-and-move-groupedfields
+  [v]
+  (let [result (zip/post-order-visit
+                 (zip/tree-zipper v)
+                 #{}
+                 [(fn [node state]
+                    (cm/match node
+                              [:groupedfield field]
+                              {:node field :state (conj state field)}
+
+                              :else nil))
+                  (fn  [node state]
+                    (when (not (empty? state))
+                      (cm/match node
+                                ["extract" & _]
+                                {:node (conj node (vec (cons "group_by" state)))}
+
+                                :else nil)))])]
+    (vec (:node result))))
+
 (defn slurp-expr->extract
   [clauses]
   (let [paging-groups (group-by paging-clause? clauses)
         paging-clauses (get paging-groups true)
-        other-clauses (get paging-groups false)]
-    (if (and (= (ffirst other-clauses) "extract") (second other-clauses))
-      (cons (vec (concat (first other-clauses) (rest other-clauses))) (vec paging-clauses))
-      clauses)))
+        other-clauses (get paging-groups false)
+        result (if (and (= (ffirst other-clauses) "extract") (second other-clauses))
+                 (cons (vec (concat (first other-clauses) (rest other-clauses)))
+                       (vec paging-clauses))
+                 clauses)]
```

Comment:
I'm still not really satisfied with this stuff hanging around. I'd like to patch it up to use zippers as well.